containers/jvm/clipper-spark/src/main/scala/ai/clipper/spark/Clipper.scala

```diff
@@ -0,0 +1,346 @@
+                           "-o", "StrictHostKeyChecking=no",
+                           "-i", s"$sshKeyPath",
+                           s"$sshUserName@$host",
+                           "mkdir", "-p", destPath)
+    val copyCommand = Seq("rsync",
+                          "-r",
+                          "-e", s"ssh -i $sshKeyPath -o StrictHostKeyChecking=no",
+                          s"$localPath/",
+                          s"$sshUserName@$host:$destPath")
+    if (!(mkdirCommand.! == 0 && copyCommand.! == 0)) {
+        throw new ModelDeploymentError(
+        "Error copying model data to Clipper host")
+    }
+  }
+
+  private[clipper] def saveSparkModel[M](sc: SparkContext,
+                                         name: String,
+                                         version: Int,
+                                         model: M,
+                                         containerClass: Class[_],
+                                         basePath: String): Unit = {
+    // Check that Spark is not running in the REPL
+    if (getReplOutputDir(sc).isDefined) {
+      throw new UnsupportedEnvironmentException(
+        "Clipper cannot deploy models from Spark Shell")
+    }
+    val modelPath = Paths.get(basePath, MODEL_DIRECTORY).toString
+    val modelType = model match {
+      case m: MLlibModel => {
+        m.save(sc, modelPath)
+        // Because I'm not sure how to do it in the type system, check that
+        // the container is of the right type
+        // TODO: this test doesn't work from the REPL
+        try {
+          containerClass.newInstance.asInstanceOf[MLlibContainer]
+        } catch {
+          case e: ClassCastException => {
+            throw new IllegalArgumentException(
+              "Error: Container must be a subclass of MLlibContainer")
+          }
+        }
+        MLlibModelType
+      }
+      case p: PipelineModel => {
+        p.save(modelPath)
+        // Because I'm not sure how to do it in the type system, check that
+        // the container is of the right type
+        try {
+          containerClass.newInstance.asInstanceOf[PipelineModelContainer]
+        } catch {
+          case e: ClassCastException => {
+            throw new IllegalArgumentException(
+              "Error: Container must be a subclass of PipelineModelContainer")
+          }
+        }
+        PipelineModelType
+      }
+      case _ =>
+        throw new IllegalArgumentException(
+          s"Illegal model type: ${model.getClass.getName}")
+    }
+    val jarPath = Paths.get(
+      containerClass.getProtectionDomain.getCodeSource.getLocation.getPath)
+    val copiedJarName = CONTAINER_JAR_FILE
+    println(s"JAR path: $jarPath")
+    System.out.flush()
+    Files.copy(jarPath, Paths.get(basePath, copiedJarName), REPLACE_EXISTING)
+    val conf =
+      ClipperContainerConf(containerClass.getName, copiedJarName, modelType)
+    getReplOutputDir(sc) match {
+      case Some(classSourceDir) => {
```

Comment:
That's true, but Scala wants an exhaustive pattern match. And I'm leaving some commented out code in there intentionally because I hope to support deploying from the REPL in the future.